Add "Manage Expenditure" related dialogs

ArmandoGraterol opened the pull request:
Examples of things to test:
- That the validation is working as expected for every field.
- That all of the states of the fields that are present on the Dapp are also present here. (Except for stuff not implemented yet like motions, of course).
- That the styles have not changed. 

![FireShot Capture 786 - Colony CDapp - localhost](https://user-images.githubusercontent.com/18473896/214378933-006e8812-005d-4a35-9430-b47c15452cdd.png)
![FireShot Capture 787 - Colony CDapp - localhost](https://user-images.githubusercontent.com/18473896/214378993-2e713122-5995-4bd3-827e-e3eb8d73af88.png)

Contributes to #58

willm30 commented on src/components/shared/Fields/InputStatus/InputStatus.tsx:
```diff
@@ -1,19 +1,26 @@
+    | string
+    | MessageDescriptor
+    | FieldError
+    | Merge<FieldError, FieldErrorsImpl<any>>
```
These types are HookForm types, no? But this isn't the HookForm status component. Also, you should be able to cast the error as a `Message` type since we control what the error message is.

ArmandoGraterol replied:
You are correct. I didn't notice that a HookForm status component already existed

willm30 commented on src/components/shared/Fields/InputStatus/InputStatus.tsx:
```diff
@@ -33,7 +40,19 @@ const InputStatus = ({
 }: Props) => {
   const { formatMessage } = useIntl();
-  const errorText = typeof error === 'object' ? formatMessage(error) : error;
+  const getErrorText = () => {
```
Feels to me like this is doing the work of the `formatText` utility we have on `master`

willm30 commented on src/components/shared/Fields/Textarea/Textarea.tsx:
```diff
@@ -81,13 +75,14 @@ const Textarea = ({
+    register,
+    formState: { errors },
+    watch,
+  } = useFormContext();
```
There is a HookForm text area component already on master.

willm30 commented on src/components/shared/Fields/Toggle/Toggle.tsx:
```diff
@@ -52,8 +54,9 @@ const Toggle = ({
 }: Props) => {
-  const [{ onChange, value }] = useField(name);
-
+  const { register, getValues } = useFormContext();
```
We also have a hook form toggle component on master.

willm30 commented on src/utils/checks/userHasRoles.ts:
```diff
@@ -0,0 +1,19 @@
+import { ColonyRole } from '@colony/colony-js';
+
+export const userHasRole = (userRoles: ColonyRole[], role: ColonyRole) =>
```
Liking these 😎

willm30 reviewed (changes requested):
Nice start @ArmandoGraterol! Requesting changes specifically to the number formatting below as I don't see that in the dapp, but also have some general comments/questions.

Long numbers change their format but it looks a little peculiar, as if it's actually becoming a decimal and therefore smaller. I suppose it's missing the `x10^8` or similar suffix? 

![numberslong](https://user-images.githubusercontent.com/64402732/214547169-be03456e-88e8-4c6b-bc34-c1b4b77e2f15.gif)

Also noticing the "(you)" below isn't in the current dapp in production. Any reason for its inclusion?

![wrognaddress](https://user-images.githubusercontent.com/64402732/214547110-df1835ef-1d92-4dde-9865-aeba798ea836.png)

ArmandoGraterol commented on src/components/common/Dialogs/CreatePaymentDialog/CreatePaymentDialog.tsx:
```diff
@@ -0,0 +1,248 @@
+  },
+});
+
+export interface FormValues {
```
Nice tip!

ArmandoGraterol commented on src/components/common/Dialogs/CreatePaymentDialog/CreatePaymentDialog.tsx:
```diff
@@ -0,0 +1,248 @@
+            <DialogForm
+              {...formState}
+              values={values}
+              setValue={setValue}
```
Is our aim to just not pass values as props that can be obtained through hooks? Jakub mentioned that we are moving out from passing `colony` as a prop, and now this. 

I'm asking this because I'm wondering if that's really worth it from a performance point of view as we would suddenly have lots of hooks requesting values that may already be available 1 level above

jakubcolony replied:
I think it depends on the situation. Here I agree with Will that it makes sense to call the hook within the consuming component rather than pass the props.
What concerns about performance do you have specifically? Will using hooks in multiple places impact it negatively? I wouldn't think it'd make a huge difference, at the end of the day whenever the value of the context changes, `DialogForm` will be re-rendered, no matter if it's getting the values via props or a hook

ArmandoGraterol replied:
Yeah, my concern was mostly thinking if it was worth it using a hook to get data that is being used anyways by the parent component. 

I did a little search and read some stuff about prop drilling vs using context, and looks like in our case performance wouldn't be a significant issue on whether we use one or the other. In general, so long as the data provided by the context is not something that is constantly changing, we are good.

That, plus the other benefits of using context in comparison with prop drilling, I would have to agree with you guys on using contexts more :)

ArmandoGraterol commented on src/components/common/Dialogs/CreatePaymentDialog/CreatePaymentDialog.tsx:
```diff
@@ -0,0 +1,248 @@
+    () => watchers?.items.filter(notNull).map((item) => item.user) || [],
+    [watchers],
+  );
+  const colonyTokens = useMemo(() => tokens?.items || [], [tokens]);
```
Makes sense to me. Thank you for the article!

I feel like what I liked the most and it's a pro to not just deliberately use `useMemo` and `useCallback` is that unless the performance of the CDapp is "screaming" at us due to expensive functions/calculations, we may do ourself a disservice by just assuming that every value/function belongs in a `useMemo/useCallback`.

ArmandoGraterol replied:
I've decided to discard these changes as it seems like I have forgotten about javascript inability to properly deal with large numbers, however, if we decide that we still want to transform input values to numbers in the feature, I still have those changes stored in my stash

ArmandoGraterol commented on src/components/common/Dialogs/CreatePaymentDialog/CreatePaymentDialogForm.tsx:
```diff
@@ -0,0 +1,682 @@
+
+// const UserAvatar = HookedUserAvatar({ fetchUser: false });
+
+const supRenderAvatar = (address: Address, item: ItemDataType<User>) => (
```
I think it's `support`, however, I agree that the name is confusing. then again this is as of now a 2-year-old name so it probably made more sense at that time. 

I'll rename it to simply `renderAvatar` or something like that.

ArmandoGraterol commented on src/components/shared/Fields/Form/HookForm.tsx:
```diff
@@ -4,21 +4,31 @@ import {
 
 const displayName = 'HookForm';
 
+export type CustomSubmitHandler<FormData> = (
```
Yes I can, thank you!

ArmandoGraterol commented on src/components/shared/Fields/Input/HookForm/InputComponent.tsx:
```diff
@@ -62,30 +63,41 @@ const HookFormInputComponent = ({
-    ref,
-  } = register(name);
+  const { control } = useFormContext();
+  const { field } = useController({
```
One of the things I changed related to the dialog forms is that the amounts we store in the form state are actually numbers instead of strings. I did this so that the types and validation we assign to number fields match the actual value (This wasn't as obvious of a problem in the Dapp because the typing was looser there).

This logic is something I extracted from [here](https://react-hook-form.com/advanced-usage#TransformandParse). The idea is that whenever the onBlur and onChange handlers are fired on the inputs/Cleavers, instead of passing the raw value string to the form state we transform them into numbers.

This should also answer the question below. 

Let me know what you think!

ArmandoGraterol replied:
I've decided to discard these changes as it seems like I have forgotten about javascript inability to properly deal with large numbers, however, if we decide that we still want to transform input values to numbers in the feature, I still have those changes stored in my stash

ArmandoGraterol commented on src/components/common/Dialogs/CreatePaymentDialog/CreatePaymentDialogForm.tsx:
```diff
@@ -0,0 +1,682 @@
+  );
+
+  return (
+    <>
```
Can you give me an example of something that could be encapsulated in this component? I feel like I understand what you are referring to but I want to make sure. For example:
```
        <div className={styles.tokenAmount}>
          <div className={styles.tokenAmountInputContainer}>
            <Input
              label={MSG.amount}
              name="amount"
              appearance={{
                theme: 'minimal',
                align: 'right',
              }}
              formattingOptions={formattingOptions}
              disabled={inputDisabled}
              dataTest="paymentAmountInput" 
            />
            {/* {networkFeeInverse &&
              customAmountError === undefined &&
              values.amount &&
              Number(values.amount) > 0 && (
                <div className={styles.networkFee}>
                  <FormattedMessage
                    {...MSG.fee}
                    values={{
                      fee: (
                        <Numeral
                          appearance={{
                            size: 'small',
                            theme: 'grey',
                          }}
                          value={
                            calculateFee(
                              values.amount,
                              networkFeeInverse,
                              getTokenDecimalsWithFallback(
                                selectedToken?.decimals,
                              ),
                            ).feesInWei
                          }
                          unit={getTokenDecimalsWithFallback(
                            selectedToken && selectedToken.decimals,
                          )}
                        />
                      ),
                      symbol: (selectedToken && selectedToken.symbol) || '???',
                    }}
                  />
                </div>
              )} */}
          </div>
          <div className={styles.tokenAmountContainer}>
            <div className={styles.tokenAmountSelect}>
              <TokenSymbolSelector
                label={MSG.token}
                tokens={colonyTokens}
                name="tokenAddress"
                elementOnly
                appearance={{ alignOptions: 'right', theme: 'grey' }}
                disabled={inputDisabled}
              />
            </div>
            {values.tokenAddress === AddressZero && (
              <div className={styles.tokenAmountUsd}>
                <EthUsd
                  // appearance={{ theme: 'grey' }}
                  value={amount}
                />
              </div>
            )}
          </div>
        </div>
```
Would you say that this part can be encapsulated further?

willm30 replied:
I generally think of it like each distinct "thing" in the form is its own component, e.g. an input or a selector, which may itself be composed of smaller components. So in this form, I see: 

A Dialog Heading (which includes the domain selector and the title "Payment")
Another domain selector (from: Root)
Recipient selector
Amount selector (including the token selector)
The text area for the annotation
And the controls at the bottom (back / create motion)

I think that's what Jakub's suggesting also, because these are the things wrapped in the `DialogSection`s.

The nice thing about this exercise is that we are going to start to see patterns of repeated abstractions (every form has a heading with the domain selector and the controls at the bottom, for example), so it provides a basis for further code reuse. I'm basically using [this guide](https://reactjs.org/docs/thinking-in-react.html) as my reference when thinking about component hierarchies / encapsulation.

willm30 replied:
To your question, I'd say this represent its own "thing" (e.g. NetworkFeeMessage): 

``` js
{/* {networkFeeInverse &&
              customAmountError === undefined &&
              values.amount &&
              Number(values.amount) > 0 && (
                <div className={styles.networkFee}>
                  <FormattedMessage
                    {...MSG.fee}
                    values={{
                      fee: (
                        <Numeral
                          appearance={{
                            size: 'small',
                            theme: 'grey',
                          }}
                          value={
                            calculateFee(
                              values.amount,
                              networkFeeInverse,
                              getTokenDecimalsWithFallback(
                                selectedToken?.decimals,
                              ),
                            ).feesInWei
                          }
                          unit={getTokenDecimalsWithFallback(
                            selectedToken && selectedToken.decimals,
                          )}
                        />
                      ),
                      symbol: (selectedToken && selectedToken.symbol) || '???',
                    }}
                  />
                </div>
              )} */}

```

This condition should really live in a named variable so its easier for others to see what it's trying to achieve: 

``` js
networkFeeInverse &&
              customAmountError === undefined &&
              values.amount &&
              Number(values.amount) > 0 
```

And then I would personally look at writing a function called `getValues` or `getMessage values` that returns the values object, so that the actual `FormattedMessage` component just looks something like:

``` js

const numeralValue = calculateFee(values.amount, networkFeeInverse, getTokenDecimalsWithFallback(selectedToken?.decimals)).feesInWei

...

<FormattedMessage
                    {...MSG.fee}
                    values={getValues(numeralValue, numeralUnit, tokenSymbol)}
```

And I think the following could probably just live inside a component called `TokenSymbol`

``` js
<div className={styles.tokenAmountContainer}>
            <div className={styles.tokenAmountSelect}>
              <TokenSymbolSelector
                label={MSG.token}
                tokens={colonyTokens}
                name="tokenAddress"
                elementOnly
                appearance={{ alignOptions: 'right', theme: 'grey' }}
                disabled={inputDisabled}
              />
            </div>
            {values.tokenAddress === AddressZero && (
              <div className={styles.tokenAmountUsd}>
                <EthUsd
                  // appearance={{ theme: 'grey' }}
                  value={amount}
                />
              </div>
            )}
          </div>

```

So the finished section might look like

``` js

<>
   <PaymentAmount>
  <NetworkFeeMessage>
  <TokenSymbol>
</>

```

I would add that in general, the intention is to make it easier to see what's happening at a glance, so it's easier for the next guy to understand what's going on, and this is often achieved through moving logic into named variables and functions (since the naming process itself is an exercise in signposting intent to team members).

ArmandoGraterol commented:
> Long numbers change their format but it looks a little peculiar, as if it's actually becoming a decimal and therefore smaller. I suppose it's missing the `x10^8` or similar suffix?

Yeah. this may be javascript shenanigans with numbers I feel like. I think I'm gonna change the transform function I implemented to transform the numbers into Decimals instead of using plain numbers.

> Also noticing the "(you)" below isn't in the current dapp in production. Any reason for its inclusion?

Fixed! It was a bug in the picker related to this conditional `walletAddress === userAddress` being `true` when both props were `undefined`.

jakubcolony reviewed (changes requested):
Good start! 

Validation seems to be working ok, one question about this case:
![image](https://user-images.githubusercontent.com/112586815/214676905-319fae5e-430a-49fd-bed6-f4232bf7dd54.png)

Is the button disabled because there's not enough funds? If so, shouldn't we show some warning regarding this?

rdig commented on src/components/common/Dialogs/ManageExpenditureDialog/ManageExpenditureDialog.tsx:
```diff
@@ -0,0 +1,135 @@
+import { getAllUserRoles } from '~redux/transformers';
+import { canAdminister, canFund } from '~utils/checks';
+
+const displayName = 'common.ColonyHome.ManageExpenditureDialog';
```
I think this was a leftover from when we re-factored all the Dialogs locations in the Dapp.

Please correct this.

rdig reviewed (approved):
Other than what the others pointed out, I'm good with this.

ArmandoGraterol commented:
@willm30 @jakubcolony Hey! All the changes should be up. I would appreciate a re-review :)

willm30 commented on src/components/shared/SingleUserPicker/SingleUserPicker.tsx:
```diff
@@ -1,9 +1,8 @@
+import { useFormContext } from 'react-hook-form';
 
-import { useField } from 'formik';
 import { AnyUser } from '~data/index';
```
`~data/index` doesn't exist in cdapp. Also, none of the message descriptors below have been updated to use new pattern.

willm30 reviewed (commented):
Validation is a bit off for me, e.g.

Negative number, no error:

![negative](https://user-images.githubusercontent.com/64402732/216017479-2fad1fa3-b6ef-463d-b058-e02a9b21f129.png)

Message not formatted:

![mesg](https://user-images.githubusercontent.com/64402732/216017686-e94410ff-6a06-4dc0-af9d-d79ba40799af.png)

Also agree wtih Jakub that it'd be good to see something to indicate the reason the form is disabled even though everything is filled in is because there aren't enough funds.

We should also ensure that when the single user picker is touched and empty the red border shows, per this section of code:
![sup](https://user-images.githubusercontent.com/64402732/216018272-5a0d966c-69c8-4f62-ab76-8b0396c40b7e.png)

jakubcolony commented on src/components/common/Dialogs/CreatePaymentDialog/NetworkFee/NetworkFee.tsx:
```diff
@@ -0,0 +1,92 @@
+              value={
+                calculateFee(
+                  values.amount,
+                  networkFeeInverse!,
```
This triggers eslint and actually looks like unsafe assertion. This prop can be undefined and you don't have any checks that would ensure it isn't

ArmandoGraterol replied:
But we do have checks, before this render:
```
  if (
    !networkFeeInverse &&
    customAmountError &&
    new Decimal(values.amount).isZero()
  ) {
    return null;
  }
```

But I'll think about implementing something clearer to eslint

jakubcolony commented on src/components/common/Dialogs/CreatePaymentDialog/TokenAmountInput/TokenAmountInput.tsx:
```diff
@@ -0,0 +1,101 @@
+          disabled={disabled}
+          dataTest="paymentAmountInput"
+        />
+        {/* <NetworkFee colony={colony} networkFeeInverse={networkFeeInverse} customAmountError={customAmountError} /> */}
```
Why is this commented out? It was showing up earlier

ArmandoGraterol replied:
Are you certain? This is a new component I introduced in the latest commits after your first review, and it should contain the code that shows the network fee. The code inside this component was also commented out before I extracted it in this component.

Why has this been commented out before and now: I think there's currently no query to get the network fee (In the Dapp you could get it like this: `const { feeInverse: networkFeeInverse } = useNetworkContracts();`).

jakubcolony replied:
It could have been in the Dapp... That would make sense...

jakubcolony reviewed (changes requested):
@ArmandoGraterol nice work splitting up that massive component!

Generally looking good, but the validation is still a bit broken. The button is disabled when both assignee and amount is filled in (presumably because the balance is too low but there isn't a message saying that), then when you "deselect" the user (see video) and tweak the amount, the button becomes active:

https://user-images.githubusercontent.com/112586815/216453888-f33d6eb1-71db-4505-8b22-4844cae76cc2.mov

ArmandoGraterol commented:
@jakubcolony Unfortunately, I can't replicate it :(

[7030ccfb-81fa-4c81-8606-87ed2ac664d7.webm](https://user-images.githubusercontent.com/18473896/216475565-51725b4d-4169-49e8-8e3d-b83c86e88ed7.webm)

Can you make sure that your branch is up-to-date? It's the only thing that I can think of  if you are not doing anything special besides what you showed in the video

willm30 reviewed (commented):
The form is now valid even though there aren't enough funds:

![valid-1](https://user-images.githubusercontent.com/64402732/216574224-5e1fc064-5b11-4852-bc14-62b96733f4b6.png)

I am expecting the form to be invalid if there aren't enough funds, and a message below the amount input informing the user of insufficient funds.

We don't show the message in the dapp below the single user picker, just the red border. From memory, the forms in the gnosis safe branch have the behaviour I'm thinking of.

![wrong](https://user-images.githubusercontent.com/64402732/216574877-43c37b31-7275-40e9-9554-29fd251f400c.png)

ArmandoGraterol commented:
> I am expecting the form to be invalid if there aren't enough funds, and a message below the amount input informing the user of insufficient funds.

Do you expect that even though getting the balance of the colonies is not something that has been implemented yet?

> We don't show the message in the dapp below the single user picker, just the red border. From memory, the forms in the gnosis safe branch have the behaviour I'm thinking of.

This was intended as an improvement, as before the picker only showed message errors if `appearance` had the `horizontal` prop:
```
            {error && appearance && appearance.direction === 'horizontal' && (
              <span className={styles.errorHorizontal}>{error}</span>
            )}
```

So I'll take it that you disagree with this change? Shouldn't we also show an error message for the picker if we can, just like we do for any other of our inputs?

jakubcolony reviewed (approved):
LGTM, didn't experience any problems with the validation, nice work! 👍 

The only little thing I noticed is the user avatar in the SingleUserPicker dropped a bit after adding the error message:
<img width="398" alt="image" src="https://user-images.githubusercontent.com/112586815/216702271-9a42ccb0-0db0-462a-b428-30495db90922.png">
That's how it is in the Dapp:
<img width="388" alt="image" src="https://user-images.githubusercontent.com/112586815/216702350-bc89a30c-0b8b-481b-9eef-0dfc91fce950.png">

ArmandoGraterol commented:
@jakubcolony 

> The only little thing I noticed is the user avatar in the SingleUserPicker dropped a bit after adding the error message.

What in the tarnation? For me, it's looking the same as in the dapp :thinking: In what OS are you? It might be some CSS OS alignment shenanigans

jakubcolony commented:
@ArmandoGraterol 
>What in the tarnation? For me, it's looking the same as in the dapp 🤔 In what OS are you? It might be some CSS OS alignment shenanigans

Chrome on MacOS. And yeah, that sounds likely. I doubt it's worth spending time on

willm30 commented on src/utils/actions.ts:
```diff
@@ -1,7 +1,7 @@
 import { UniqueActionType } from '~redux';
 
-export type ActionTransformFnType = (
+export type ActionTransformFnType = () => (
```
I don't think this is correct. Transform gets called with the action as an argument.

ArmandoGraterol replied:
I remember that at some point typescript was giving me an error because of `useCallback(pipe...)` but we are removing `useCallback` nevertheless

willm30 reviewed (approved):
Nice work Armando, approving but would appreciate it if you could look over my notes.

> This was intended as an improvement,

Ok no prob. In that case, perhaps update the message so it's specific to the user picker, e.g. "Please enter an address"

ArmandoGraterol commented on src/components/common/Dialogs/CreatePaymentDialog/CreatePaymentDialogForm.tsx:
```diff
@@ -0,0 +1,300 @@
+  /*
+   * Custom error state tracking
+   */
+  const [customAmountError] = useState<
```
Yeah I agree :ok_hand: I feel like with what we know of Yup and Hook Form we will hopefully not need custom errors like this one (I have removed some of them already with this port)

ArmandoGraterol commented on src/components/common/Dialogs/CreatePaymentDialog/CreatePaymentDialogForm.tsx:
```diff
@@ -0,0 +1,300 @@
+  const inputDisabled = !canMakePayment || onlyForceAction || isSubmitting;
+
+  const formattedData = useMemo(
+    () => verifiedUsers.map((user) => ({ ...user, id: user.walletAddress })),
```
Good idea!

ArmandoGraterol commented on src/components/common/Dialogs/CreatePaymentDialog/DomainFundSelector/DomainFundSelector.tsx:
```diff
@@ -0,0 +1,207 @@
+      ? Id.RootDomain
+      : filteredDomainId;
+
+  const domainId = values.domainId
```
You are correct, this is old code

ArmandoGraterol commented on src/components/common/Dialogs/CreatePaymentDialog/TokenAmountInput/TokenAmountInput.tsx:
```diff
@@ -0,0 +1,101 @@
+  const selectedToken = colonyTokens.find(
+    (token) => token?.tokenAddress === values.tokenAddress,
+  );
+  const formattingOptions = useMemo(
```
I'm gonna have to cite one of the greatest thinker from our generation, Will from 2 months ago:

> * @NOTE: If formattingOptions is not either memoized or defined outside of the ancestor Input component,
> * it will cause Cleave to be re-mounted and thus lose its state and focus.


If I don't use `useMemo` for `formattingOptions`, everytime I type something I lose focus :(

ArmandoGraterol commented on src/components/shared/Fields/Input/HookForm/Input.tsx:
```diff
@@ -99,6 +99,7 @@ const HookFormInput = ({
           name={name}
           placeholder={formatText(placeholder, placeholderValues)}
           inputValueLength={inputValue?.length || 0}
+          value={inputValue}
```
At some point, I did this to fix some weird behavior the input had, but it seems like it's not happening anymore :thinking:
